backend/drm/drm.c

```diff
@@ -296,54 +296,65 @@ static void wlr_drm_output_transform(struct wlr_output_state *output,
 	output->wlr_output->transform = transform;
 }
 
-static void wlr_drm_cursor_bo_update(struct wlr_output_state *output,
-		uint32_t width, uint32_t height) {
-	if (output->cursor_width == width && output->cursor_height == height) {
-		return;
-	}
-	wlr_log(L_DEBUG, "Allocating new cursor bos");
-	struct wlr_backend_state *state =
-		wl_container_of(output->renderer, state, renderer);
-	for (size_t i = 0; i < 2; ++i) {
-		output->cursor_bo[i] = gbm_bo_create(state->renderer.gbm,
-				width, height, GBM_FORMAT_ARGB8888,
-				GBM_BO_USE_CURSOR | GBM_BO_USE_WRITE);
-		if (!output->cursor_bo[i]) {
-			wlr_log(L_ERROR, "Failed to create cursor bo");
-			return;
-		}
-		if (!get_fb_for_bo(state->fd, output->cursor_bo[i])) {
-			wlr_log(L_ERROR, "Failed to create cursor fb");
-			return;
-		}
-	}
-}
-
 static bool wlr_drm_output_set_cursor(struct wlr_output_state *output,
 		const uint8_t *buf, int32_t stride, uint32_t width, uint32_t height) {
-	struct wlr_backend_state *state =
-		wl_container_of(output->renderer, state, renderer);
+	struct wlr_backend_state *state = wl_container_of(output->renderer, state, renderer);
 	if (!buf) {
 		drmModeSetCursor(state->fd, output->crtc, 0, 0, 0);
 		return true;
 	}
-	wlr_drm_cursor_bo_update(output, width, height);
+
+	uint64_t bo_width, bo_height;
+	int ret;
+
+	ret = drmGetCap(state->fd, DRM_CAP_CURSOR_WIDTH, &bo_width);
+	bo_width = ret ? 64 : bo_width;
+	ret = drmGetCap(state->fd, DRM_CAP_CURSOR_HEIGHT, &bo_height);
+	bo_height = ret ? 64 : bo_height;
+
+	if (width > bo_width || height > bo_width) {
+		wlr_log(L_INFO, "Cursor too large");
+		return false;
+	}
+
+	for (int i = 0; i < 2; ++i) {
+		if (output->cursor_bo[i]) {
+			continue;
+		}
+
+		output->cursor_bo[i] = gbm_bo_create(state->renderer.gbm, bo_width, bo_height,
+			GBM_FORMAT_RGBA8888, GBM_BO_USE_CURSOR | GBM_BO_USE_WRITE);
```

Comment:
Yeah, the cat is in RBGA format, so it would be reading the colours incorrectly as ARBG. I think only ARGB8888 and XRBG8888 are the only guaranteed formats to be available, so I guess some code to convert the pixel format is needed, or the cat could be changed.